### ISSUE TITLE
web2 submitRequest result does not have content-length

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2284]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)| @ghazel reported|
|-|-|
|Trac ID|trac#2284|
|Type|defect|
|Created|2006-12-08 18:27:13Z|

The response object given as a result of HTTPClientProtocol.submitRequest does not have a reference to content-length.

The fix seems to be simple:

```
Index: client/http.py
===================================================================
--- client/http.py      (revision 18823)
+++ client/http.py      (working copy)
@@ -164,7 +164,7 @@

     ## FIXME: Actually creates Response, function is badly named!
     def createRequest(self):
-        self.stream = stream_mod.ProducerStream()
+        self.stream = stream_mod.ProducerStream(self.length)
         self.response = http.Response(self.code, self.inHeaders, self.stream)
         self.stream.registerProducer(self, True)
```

Attachments:

* [fix.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2ee/2eee5f04b3a9f4ae8e5232f6fcd58d93238c3ece/653b8a47b62a3d2b9e03df768e3b54c79b513045.patch) (7517 bytes) - added by jknight on 2006-12-27 20:29:20Z - 
* [fix.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2ee/2eee5f04b3a9f4ae8e5232f6fcd58d93238c3ece/68c96aa606c84b48e8209209f1185c10a5b0a665.patch) (16295 bytes) - added by jknight on 2007-01-01 04:44:34Z - 2nd try.

<details><summary>Searchable metadata</summary>

```
trac-id__2284 2284
type__defect defect
reporter__ghazel ghazel
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1165602433000000 1165602433000000
changetime__1167694890000000 1167694890000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_1"></a>|@jyknight removed owner|
|-|-|

Yep. The attached patch fixes this and [#2302](https://github.com/twisted/twisted/issues/2302), as well as correcting some bogus tests which were non-functional due to test bugs.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_2"></a>|@jerub set owner to @jyknight|
|-|-|

Please don't use this idiom for conditional expressions. Please express this another way.
```
['Connection: '+ (second and 'close' or 'Keep-Alive')]
```

Does this change mean that the resp.getAllRawHeaders() no longer returns a key of content-length? That's what the tests seem to indicate.

test_userHeaders, test_sentHeadPipelined, test_prematurePipelining, test_streamedUpload, test_sentHead, test_serverDoesntSendConnectionClose are all modified or created test methods but do not have docstrings. Please add docstrings to these methods.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_3"></a>|@jyknight commented|
|-|-|

Replying to [jerub](#note_2):
> Does this change mean that the resp.getAllRawHeaders() no longer returns a key of content-length? That's what the tests seem to indicate.

It never did. The tests were not testing what they thought they were.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_4"></a>|@jerub commented|
|-|-|

Ah, good to know.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_5"></a>|@jyknight commented|
|-|-|

Okay, I figured out what it should do: Content-Length isn't actually a connection header for HEAD. 

So here's a better patch, which makes Content-Length a normal header for HEAD requests. This both fixes the bug, and makes it available in response.headers to a HEAD request. 

Also fixed Jerub's objections.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight set owner to @jerub|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @jyknight<br>@exarkun set status to `new`|
|-|-|

If I understand correctly, this changes the client code to notice when it has made a HEAD request and in that case, makes certain to preserve the information conveyed by Content-Length in the response so that it is available to the application along with the other headers.  This is as opposed to the alternate case of a non-HEAD request, where the Content-Length header is consumed by the client code and then re-exposed to application code via the stream.length attribute.

If that's wrong, I might need some help understanding what's going on.

If I've understood that correctly, though, then I have only these fairly minor comments about the change:

  * in splitConnectionHeaders, is ```if self.length is not None:``` certain to correctly determine that a HEAD request was made?  It seems like a somewhat round-about way of checking for this case.  The comment implies that this is what's going on, but if there isn't a more obvious test to be made, it might be worth making this invariant explicit.
  * ```del connHeaderNames[0]``` might be clearer as ```connHeaderNames.remove('content-length')```.
  * splitConnectionHeaders could use a docstring perhaps
  * trailing whitespace all over the place, oh my.  At least some of it is cleaned up in this branch.

Also, I like the refactoring in the test module.  The only thing that I really definitely think should be fixed before merging is the splitConnectionHeaders docstring.

**Comment by twisted-trac** [comment]
|<img alt="foom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|foom set status to `closed`|
|-|-|

(In [19180]) Fix content-length issues in HTTP client:
Propagate content-length into stream.length. (Fixes [#2284](https://github.com/twisted/twisted/issues/2284))
Fix HEAD requests with content-length. (Fixes [#2302](https://github.com/twisted/twisted/issues/2302))
Also, refactor and fix the test_client tests and add some docstrings.

Reporter: ghazel
Author: jknight
Reviewer: jerub, exarkun

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
